### PR TITLE
new parameterupdater use paddle pserver cclient of go

### PR DESCRIPTION
**Pull request by jacquesqiao**
refs: 
issue: https://github.com/PaddlePaddle/Paddle/issues/2274
old pr: https://github.com/PaddlePaddle/Paddle/pull/2330

测试方法：
1，编译。
2，编译完成之后运行 go run ./go/cmd/pserver/pserver.go -port 3000 会在本地启动一个parameter server，监听3000端口。
3，python paddle/go/pserver/cclient/test/test_train.py

增加了一个简单的梯度下降optimizer，已经可以正常训练和收敛了

### CODE REVIEW

File: paddle/trainer/NewRemoteParameterUpdater.h [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  virtual bool finishPass();
+
+  int parameterSize() { return (int)parameters_.size(); }
+
+  /**
+   * init parameter of paddle pserver cclient.
+   * @param new_paras
+   * @param type
+   */
+  void initNewParameter(paddle_parameter**& new_paras, ParameterType type) {
```
Comment (dzhwinter) [inline paddle/trainer/NewRemoteParameterUpdater.h]: 这个引用去掉是不是比较好？这里只是 *type，指针拷贝没有代价。这样写风格有点怪~

File: go/pserver/cclient/test/main.c [hunk 2]
```diff
@@ -41,30 +43,31 @@ int main() {
 
   unsigned char content[] = {0x00, 0x11, 0x22};
   paddle_gradient grads[2] = {
-      {"param_a", PADDLE_ELEMENT_TYPE_INT32, content, 3},
-      {"param_b", PADDLE_ELEMENT_TYPE_FLOAT32, content, 3}};
+      {"param_a", PADDLE_ELEMENT_TYPE_FLOAT32, content, 3},
```
Comment (dzhwinter) [inline go/pserver/cclient/test/main.c]: yes! I remember there is a bug here when I test the pserver program

File: paddle/trainer/NewRemoteParameterUpdater.h [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    new_paras =
+        (paddle_parameter**)malloc(sizeof(paddle_parameter*) * parameterSize());
+    for (int i = 0; i < parameterSize(); ++i) {
+      new_paras[i] = (paddle_parameter*)malloc(sizeof(paddle_parameter));
```
Comment (dzhwinter) [inline paddle/trainer/NewRemoteParameterUpdater.h]: 没有找到对应的free。
另外，
```C
//cclient.go
static inline void paddle_release_param(paddle_parameter* param) 
```
释放parameter内存不应该由cclient负责吧？ @helinwang ，既然由remoteUpdater申请内存，释放空间最好还是在同一个scope内，否则容易出错。
我觉得这里用智能指针更好一些？

File: paddle/trainer/NewRemoteParameterUpdater.h [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+      memset(new_paras[i], 0, sizeof(paddle_parameter));
+    }
+
+    for (int i = 0; i < parameterSize(); ++i) {
+      ParameterPtr para = parameters_[i];
+      new_paras[i]->content_len = 10;
+      new_paras[i]->element_type = PADDLE_ELEMENT_TYPE_FLOAT32;
+      new_paras[i]->name = (char*)para->getName().c_str();
+      new_paras[i]->content =
+          (unsigned char*)(para->getBuf(type).get()->getData());
```
Comment (dzhwinter) [inline paddle/trainer/NewRemoteParameterUpdater.h]: just a suggestion, it's up to you, do we need to unify the variable abbrevtion？ helin used to named parameter with `param`, this part named with para. How do you think? Both is fine to me.

File: paddle/trainer/NewRemoteParameterUpdater.h [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+/**
+ * New remote parameter updater for dense parameters that use cclient of go.
+ */
+class NewRemoteParameterUpdater : public ParameterUpdater {
+public:
+  NewRemoteParameterUpdater(const OptimizationConfig& config,
+                            const std::string pserverSpec);
+  ~NewRemoteParameterUpdater() {
+    if (newGradients_) {
```
Comment (dzhwinter) [inline paddle/trainer/NewRemoteParameterUpdater.h]: I can't understand this condition, and we need to free the newGradient after release client, aren't we?

**Comment by helinwang** [comment]
@jacquesqiao In future it's better to use the same PR, so reviewer can track the comment / reply. You can just force push to the PR branch if you rebased / did a merge.

File: paddle/trainer/NewRemoteParameterUpdater.h [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    new_paras =
+        (paddle_parameter**)malloc(sizeof(paddle_parameter*) * parameterSize());
+    for (int i = 0; i < parameterSize(); ++i) {
+      new_paras[i] = (paddle_parameter*)malloc(sizeof(paddle_parameter));
+      memset(new_paras[i], 0, sizeof(paddle_parameter));
+    }
+
+    for (int i = 0; i < parameterSize(); ++i) {
+      ParameterPtr para = parameters_[i];
+      new_paras[i]->content_len = 10;
```
Comment (helinwang) [inline paddle/trainer/NewRemoteParameterUpdater.h]: I think having 10 here is not correct.

File: paddle/trainer/NewRemoteParameterUpdater.h (hunk 3 above)
Comment (helinwang) [inline paddle/trainer/NewRemoteParameterUpdater.h]: 对的，parameter内存不由cclient拥有。

Comment (helinwang) [inline paddle/trainer/NewRemoteParameterUpdater.h]: remoteUpdater destruct的时候需要释放对应成员变量的内存。

File: paddle/trainer/NewRemoteParameterUpdater.h [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+      new_paras[i]->element_type = PADDLE_ELEMENT_TYPE_FLOAT32;
```
Comment (typhoonzero) [inline paddle/trainer/NewRemoteParameterUpdater.h]: Need a comment that type is defined in "Vector.h" with `typedef VectorT<real> Vector;`, so parameter element type is always `real`

File: paddle/trainer/NewRemoteParameterUpdater.cpp [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+#include "NewRemoteParameterUpdater.h"
+#include "Trainer.h"
+#include "paddle/utils/Stat.h"
+
+DECLARE_int32(trainer_id);
+DECLARE_string(save_dir);
+
+namespace paddle {
+NewRemoteParameterUpdater::NewRemoteParameterUpdater(
+    const OptimizationConfig &config, const std::string pserverSpec)
```
Comment (dzhwinter) [inline paddle/trainer/NewRemoteParameterUpdater.cpp]: In fact, one OptimizationConfig is not enough, it needs more kinds optimizers when training a complicated network. e.g. some part use `adagrad` optimizer, some part use `Ftrl` optimizer https://arxiv.org/abs/1606.07792, which is I mentioned in last week. we can polish this part after we combining these modules together.

File: paddle/trainer/NewRemoteParameterUpdater.h [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+  ~NewRemoteParameterUpdater() {
+    if (newGradients_) {
+      paddle_pserver_client_release(parameterClient_);
+    }
+  }
+
+  /**
+   * initialize the internal parameter client and itself.
+   */
+  virtual void init(const std::vector<ParameterPtr>& parameters);
```
Comment (dzhwinter) [inline paddle/trainer/NewRemoteParameterUpdater.h]: we will change the style to google c++ style, it is the most popular style in c++ opensource world, why not?

Comment (dzhwinter) [inline paddle/trainer/NewRemoteParameterUpdater.h]: emm... Do we need to unify the Go style and c++ style in naming function? It seems that Go style function naming convention is better.  How do you think?

File: paddle/trainer/NewRemoteParameterUpdater.h (hunk 5 above)
Comment (jacquesqiao) [inline paddle/trainer/NewRemoteParameterUpdater.h]: done

File: paddle/trainer/NewRemoteParameterUpdater.h (hunk 1 above)
Comment (jacquesqiao) [inline paddle/trainer/NewRemoteParameterUpdater.h]: done

File: paddle/trainer/NewRemoteParameterUpdater.h (hunk 3 above)
Comment (jacquesqiao) [inline paddle/trainer/NewRemoteParameterUpdater.h]: done

File: paddle/trainer/NewRemoteParameterUpdater.h (hunk 4 above)
Comment (jacquesqiao) [inline paddle/trainer/NewRemoteParameterUpdater.h]: done

File: go/pserver/cclient/test/main.c (hunk 2 above)
Comment (jacquesqiao) [inline go/pserver/cclient/test/main.c]: done

File: paddle/trainer/NewRemoteParameterUpdater.cpp [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+  LOG(INFO) << "NewRemoteParameterUpdater init in";
+
+  for (auto &para : parameters_) {
+    para->getBuf(PARAMETER_VALUE)->zeroMem();
+    para->getBuf(PARAMETER_GRADIENT)->zeroMem();
+  }
+
+  // create parameter server client.
+  parameterClient_ =
+      paddle_new_pserver_client((char *)pserverSpec_.c_str(), FLAGS_trainer_id);
```
Comment (jacquesqiao) [inline paddle/trainer/NewRemoteParameterUpdater.cpp]: done

File: paddle/trainer/NewRemoteParameterUpdater.cpp [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+  if (paddle_begin_init_params(parameterClient_)) {
+    LOG(INFO) << "paddle_begin_init_params start";
+    for (int i = 0; i < parameterSize(); ++i) {
+      paddle_init_param(parameterClient_, *newParameters_[i], NULL, 0);
+    }
+    paddle_finish_init_params(parameterClient_);
+    LOG(INFO) << "paddle_begin_init_params done";
+  } else {
+    paddle_get_params(
+        parameterClient_, names_, newParameters_, (int)parameters_.size());
```
Comment (jacquesqiao) [inline paddle/trainer/NewRemoteParameterUpdater.cpp]: done

File: paddle/trainer/NewRemoteParameterUpdater.h (hunk 6 above)
Comment (jacquesqiao) [inline paddle/trainer/NewRemoteParameterUpdater.h]: done, 笔误

File: paddle/trainer/NewRemoteParameterUpdater.h [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+  /**
+   * work need to do after finishBatch
+   */
+  virtual void updateImpl(Parameter* para);
+
+protected:
+  /// internal parameter client object for exchanging data with pserver
+  client parameterClient_ = -1;
+  /// the parameters for new pserver client
+  paddle_parameter** newParameters_;
```
Comment (jacquesqiao) [inline paddle/trainer/NewRemoteParameterUpdater.h]: done

File: paddle/trainer/NewRemoteParameterUpdater.h (hunk 7 above)
Comment (jacquesqiao) [inline paddle/trainer/NewRemoteParameterUpdater.h]: done

File: paddle/trainer/NewRemoteParameterUpdater.h (hunk 3 above)
Comment (jacquesqiao) [inline paddle/trainer/NewRemoteParameterUpdater.h]: 目前parameterUpdater只分配了几个指针数组，而没有分配具体的content buffer，所以只负责free自己分配的几个指针成员

File: paddle/trainer/NewRemoteParameterUpdater.cpp (hunk 8 above)
Comment (jacquesqiao) [inline paddle/trainer/NewRemoteParameterUpdater.cpp]: done, send the old parameter config to go cclient for now.

File: go/pserver/service.go [hunk 13]
```diff
@@ -29,6 +29,10 @@ type Parameter struct {
 	Content     []byte
 }
 
+func (p *Parameter) toString() {
```
Comment (jacquesqiao) [inline go/pserver/service.go]: done，removed

File: go/pserver/optimizer.c [hunk 14]
```diff
@@ -32,7 +32,14 @@ int update_SGD(void* optimizer,
                const void* gradient,
                int num_bytes) {
   SGD_optimizer* o = (SGD_optimizer*)optimizer;
-  // TODO
+  // TODO(a simple SGD implement)
```
Comment (jacquesqiao) [inline go/pserver/optimizer.c]: done, removed

File: go/pserver/cclient/test/test_cclient.c [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+  print_parameter(param_ptrs[1]);
+
+  paddle_release_param(param_ptrs[0]);
+  paddle_release_param(param_ptrs[1]);
+
+  if (paddle_save_model(c, "/tmp/") != 0) {
+    fail();
+  }
+
+  printf("test success!\n");
```
Comment (jacquesqiao) [inline go/pserver/cclient/test/test_cclient.c]: done, removed

File: paddle/trainer/NewRemoteParameterUpdater.h (hunk 9 above)
Comment (jacquesqiao) [inline paddle/trainer/NewRemoteParameterUpdater.h]: The function interface like `init`, `startBatch`of NewRemoteParameterUpdater is inherited from parent class `ParameterUpdater` that use lowercase as the first letter. So this function name has to maintain the style now, but I also agree that we use Google C++ style for C++ in the future.

**Comment by typhoonzero** [comment]
Need to resolv conflicts before merging!

**Pull request merged**